alts/src/main/java/io/grpc/alts/ComputeEngineChannelCredentials.java

```diff
@@ -66,10 +71,24 @@ public static ChannelCredentials create() {
     } catch (SSLException e) {
       throw new RuntimeException(e);
     }
+    Attributes.Key<String> clusterNameAttrKey = null;
+    try {
+      Class<?> klass = Class.forName("io.grpc.xds.InternalXdsAttributes");
+      clusterNameAttrKey = (Attributes.Key<String>) klass.getField("ATTR_CLUSTER_NAME").get(null);
+    } catch (ClassNotFoundException e) {
+      logger.log(
+        Level.FINE, "Unable to load xDS endpoint cluster name key, this may be expected", e);
+    } catch (NoSuchFieldException e) {
+      logger.log(
+        Level.FINE, "Unable to load xDS endpoint cluster name key, this may be expected", e);
+    } catch (IllegalAccessException e) {
+      logger.log(
+        Level.FINE, "Unable to load xDS endpoint cluster name key, this may be expected", e);
+    }
     return new GoogleDefaultProtocolNegotiatorFactory(
         /* targetServiceAccounts= */ ImmutableList.<String>of(),
         SharedResourcePool.forResource(HandshakerServiceChannel.SHARED_HANDSHAKER_CHANNEL),
         sslContext,
-        null);
+        clusterNameAttrKey);
```

Comment:
Why we need cluster name during protocol negotiation?

Reply:
This is for using xDS in DirectPath. In order to be capable of falling back to CFE, Traffic Director uses a special cluster resource named "google-cfe" and sends it to the client along with other cluster resources. Instead of using EDS to obtain backend addresses for normal xDS clusters, If using the "google-cfe" cluster, the gRPC client will do a DNS resolution to get the CFE addresses. Since we would need different handshake protocols for talking to DirectPath backends (ALTS) and talking to CFE (TLS), the trick we do is to attach the cluster name to addresses so that GoogleDefault protocol negotiator is able to decide which handler to use.

You can find more details in go/grpc-xds-directpath-c2p-fallback.